File: NC-150/07_Trees/57_kth_smallest_element_in_bst.py
"""
LeetCode 230: Kth Smallest Element in a BST
https://leetcode.com/problems/kth-smallest-element-in-a-bst/

Given the root of a binary search tree, and an integer k, return the kth 
smallest value (1-indexed) of all the values of the nodes in the tree.

Key Insight: Inorder traversal of BST gives elements in sorted order

Time Complexity: O(H + k) where H is tree height
Space Complexity: O(H) for recursion/stack
"""

from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    # Solution 1: Iterative Inorder Traversal (Optimal - stops early)
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        stack = []
        
        while root or stack:
            # Go to leftmost node
            while root:
                stack.append(root)
                root = root.left
            
            root = stack.pop()
            
            # Found kth smallest
            k -= 1
            if k == 0:
                return root.val
            
            root = root.right
        
        return -1  # k is larger than tree size
    
    # Solution 2: Recursive with generator (Pythonic and efficient)
    def kthSmallestGenerator(self, root: Optional[TreeNode], k: int) -> int:
        def inorder(node):
            if node:
                yield from inorder(node.left)
                yield node.val
                yield from inorder(node.right)
        
        gen = inorder(root)
        for _ in range(k):
            result = next(gen)
        return result
    
    # Solution 3: Recursive collecting all elements
    def kthSmallestRecursive(self, root: Optional[TreeNode], k: int) -> int:
        def inorder(node):
            if not node:
                return []
            return inorder(node.left) + [node.val] + inorder(node.right)
        
        return inorder(root)[k - 1]
    
    # Solution 4: Optimized recursive (stops early)
    def kthSmallestOptimized(self, root: Optional[TreeNode], k: int) -> int:
        self.count = 0
        self.result = 0
        
        def inorder(node):
            if not node or self.count >= k:
                return
            
            inorder(node.left)
            
            self.count += 1
            if self.count == k:
                self.result = node.val
                return
            
            inorder(node.right)
        
        inorder(root)
        return self.result


# Test
if __name__ == "__main__":
    # Test case 1: [3,1,4,null,2], k = 1 -> 1
    root1 = TreeNode(3)
    root1.left = TreeNode(1)
    root1.right = TreeNode(4)
    root1.left.right = TreeNode(2)
    
    # Test case 2: [5,3,6,2,4,null,null,1], k = 3 -> 3
    root2 = TreeNode(5)
    root2.left = TreeNode(3)
    root2.right = TreeNode(6)
    root2.left.left = TreeNode(2)
    root2.left.right = TreeNode(4)
    root2.left.left.left = TreeNode(1)
    
    sol = Solution()
    print(f"Test 1 (k=1): {sol.kthSmallest(root1, 1)}")           # 1
    print(f"Test 2 (k=3): {sol.kthSmallest(root2, 3)}")           # 3
    print(f"Test 2 (k=3, Generator): {sol.kthSmallestGenerator(root2, 3)}")  # 3
    print(f"Test 2 (k=3, Optimized): {sol.kthSmallestOptimized(root2, 3)}")  # 3

File: NC-150/07_Trees/57_kth_smallest_element_in_bst.cpp
/*
 * LeetCode 230: Kth Smallest Element in a BST
 * https://leetcode.com/problems/kth-smallest-element-in-a-bst/
 * 
 * Given the root of a binary search tree, and an integer k, return the kth 
 * smallest value (1-indexed) of all the values of the nodes in the tree.
 * 
 * Key Insight: Inorder traversal of BST gives elements in sorted order
 * 
 * Time Complexity: O(H + k) where H is tree height
 * Space Complexity: O(H) for recursion/stack
 */

#include <iostream>
#include <stack>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    // Solution 1: Iterative Inorder Traversal (Optimal - stops early)
    int kthSmallest(TreeNode* root, int k) {
        stack<TreeNode*> stk;
        
        while (root || !stk.empty()) {
            // Go to leftmost node
            while (root) {
                stk.push(root);
                root = root->left;
            }
            
            root = stk.top();
            stk.pop();
            
            // Found kth smallest
            if (--k == 0) {
                return root->val;
            }
            
            root = root->right;
        }
        
        return -1;  // k is larger than tree size
    }
    
    // Solution 2: Recursive Inorder (collects all elements first)
    int kthSmallestRecursive(TreeNode* root, int k) {
        vector<int> sorted;
        inorder(root, sorted);
        return sorted[k - 1];
    }
    
    // Solution 3: Optimized Recursive (stops early)
    int kthSmallestOptimized(TreeNode* root, int k) {
        int count = 0;
        int result = 0;
        inorderOptimized(root, k, count, result);
        return result;
    }
    
private:
    void inorder(TreeNode* node, vector<int>& sorted) {
        if (!node) return;
        inorder(node->left, sorted);
        sorted.push_back(node->val);
        inorder(node->right, sorted);
    }
    
    void inorderOptimized(TreeNode* node, int k, int& count, int& result) {
        if (!node) return;
        
        inorderOptimized(node->left, k, count, result);
        
        count++;
        if (count == k) {
            result = node->val;
            return;
        }
        
        inorderOptimized(node->right, k, count, result);
    }
};

int main() {
    // Test case 1: [3,1,4,null,2], k = 1 -> 1
    TreeNode* root1 = new TreeNode(3);
    root1->left = new TreeNode(1);
    root1->right = new TreeNode(4);
    root1->left->right = new TreeNode(2);
    
    // Test case 2: [5,3,6,2,4,null,null,1], k = 3 -> 3
    TreeNode* root2 = new TreeNode(5);
    root2->left = new TreeNode(3);
    root2->right = new TreeNode(6);
    root2->left->left = new TreeNode(2);
    root2->left->right = new TreeNode(4);
    root2->left->left->left = new TreeNode(1);
    
    Solution sol;
    cout << "Test 1 (k=1): " << sol.kthSmallest(root1, 1) << endl;  // 1
    cout << "Test 2 (k=3): " << sol.kthSmallest(root2, 3) << endl;  // 3
    cout << "Test 2 (k=3, Optimized): " << sol.kthSmallestOptimized(root2, 3) << endl;  // 3
    
    return 0;
}

